File: src/main/java/com/anod/appwatcher/utils/SettingsActionBarActivity.java
package com.anod.appwatcher.utils;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.anod.appwatcher.ListExportActivity;
import com.anod.appwatcher.R;
import com.anod.appwatcher.backup.ExportTask;
import com.anod.appwatcher.backup.GDriveBackup;
import com.anod.appwatcher.backup.ListExportManager;

import org.acra.ACRA;
import org.acra.ErrorReporter;

import java.util.ArrayList;
import java.util.List;

import de.psdev.licensesdialog.LicensesDialog;

abstract public class SettingsActionBarActivity extends ActionBarActivity implements AdapterView.OnItemClickListener {
    protected ListView mListView;

    public static class Preference {
        final int title;
        final int layout;
        public Preference(int title, int layout) {
            this.title = title;
            this.layout = layout;
        }
    }
    public static class Category extends Preference {
        public Category(int title) {
            super(title, R.layout.preference_category);
        }
    }

    public static class Item extends Preference {
        final int action;
        final int summaryRes;
        public String summary;
        public int widget;
        public boolean enabled = true;

        public Item(int title, int summaryRes, int action) {
            super(title, R.layout.preference_holo);
            this.summaryRes = summaryRes;
            this.action = action;
        }
    }

    static class PreferenceAdapter extends ArrayAdapter<Preference> {

        private final SettingsActionBarActivity mActivity;
        private final LayoutInflater mInflater;

        public PreferenceAdapter(SettingsActionBarActivity activity, List<Preference> objects) {
            super(activity, 0, objects);
            mActivity = activity;
            mInflater = (LayoutInflater)mActivity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }

        @Override
        public int getItemViewType(int position) {
            Preference pref = getItem(position);
            if (pref instanceof Category) {
                return 0;
            }
            return 1;
        }

        @Override
        public int getViewTypeCount() {
            return 2;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final Preference pref = getItem(position);

            View view;
            if (convertView == null) {
                view = mInflater.inflate(pref.layout, parent, false);
            } else {
                view = convertView;
            }

            TextView title = (TextView) view.findViewById(android.R.id.title);
            title.setText(pref.title);

            if (pref instanceof Item) {
                Item item = (Item) pref;
                View icon = view.findViewById(android.R.id.icon);
                icon.setVisibility(View.GONE);

                TextView summary = (TextView) view.findViewById(android.R.id.summary);
                if (item.summaryRes > 0) {
                    summary.setText(item.summaryRes);
                } else if (!TextUtils.isEmpty(item.summary)) {
                    summary.setText(item.summary);
                }

                final ViewGroup widgetFrame = (ViewGroup) view.findViewById(android.R.id.widget_frame);
                if (item.widget > 0) {
                    mInflater.inflate(item.widget, widgetFrame);
                } else {
                    widgetFrame.setVisibility(View.GONE);
                }
            }

            return view;
        }

        @Override
        public boolean isEnabled(int position) {
            final Preference pref = getItem(position);
            if(pref instanceof Category) {
                return false;
            }
            if (pref instanceof Item) {
                return ((Item) pref).enabled;
            }
            return true;
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        supportRequestWindowFeature(Window.FEATURE_INDETERMINATE_PROGRESS);
        setContentView(R.layout.preference_list_content);

        init();

        ArrayList<Preference> preferences = getPreferenceItems();

        mListView = (ListView) findViewById(android.R.id.list);
        mListView.setEmptyView(findViewById(android.R.id.empty));
        mListView.setAdapter(new PreferenceAdapter(this, preferences));
        mListView.setOnItemClickListener(this);
    }

    protected abstract void init();
    protected abstract ArrayList<Preference> getPreferenceItems();
    protected abstract void onPreferenceItemClick(int action, Item pref);
    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        Preference pref = (Preference) mListView.getItemAtPosition(position);
        if (pref instanceof Item) {
            int action = ((Item) pref).action;
            onPreferenceItemClick(action, (Item) pref);
        }
    }
}

File: framework/src/main/kotlin/info/anodsplace/framework/app/SettingsActionBarActivity.kt
package info.anodsplace.framework.app

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.annotation.StringRes
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.*
import info.anodsplace.framework.view.MenuItemAnimation
import info.anodsplace.framework.R

abstract class SettingsActionBarActivity : ToolbarActivity(), AdapterView.OnItemClickListener {
    private val listView: ListView by lazy { findViewById<View>(android.R.id.list) as ListView }
    private val refreshAnim: MenuItemAnimation by lazy { MenuItemAnimation(this, R.anim.rotate) }
    private val preferenceAdapter: PreferenceAdapter by lazy { PreferenceAdapter(this, createPreferenceItems()) }

    var viewTypeCount = 4

    open class Preference(@StringRes val title: Int, @LayoutRes val layout: Int, val viewType: Int)
    class Category(@StringRes title: Int) : Preference(title, R.layout.preference_category, 0)

    open class Item(@StringRes title: Int, @StringRes var summaryRes: Int, internal val action: Int, val widget: Int, viewType: Int)
        : Preference(title, R.layout.preference_holo, viewType) {

        constructor(@StringRes title: Int, @StringRes summaryRes: Int, action: Int, viewType: Int)
                : this(title, summaryRes, action, 0, viewType)

        var summary = ""
        var enabled = true
    }

    class TextItem(@StringRes title: Int, @StringRes summaryRes: Int, action: Int)
        : Item(title, summaryRes, action, 0, 1)

    open class ToggleItem(@StringRes title: Int, @StringRes summaryRes: Int, action: Int, widget: Int, var checked: Boolean, viewType: Int)
        : Item(title, summaryRes, action, widget, viewType) {

        fun toggle() {
            this.checked = !this.checked
        }
    }

    class CheckboxItem(@StringRes title: Int, @StringRes summaryRes: Int, action: Int, checked: Boolean = false)
        : ToggleItem(title, summaryRes, action, R.layout.preference_widget_checkbox, checked, 2)

    class SwitchItem(@StringRes title: Int, @StringRes summaryRes: Int, action: Int, checked: Boolean = false)
        : ToggleItem(title, summaryRes, action, R.layout.preference_widget_switch, checked, 3)

    internal class PreferenceAdapter(activity: SettingsActionBarActivity, objects: List<Preference>) : ArrayAdapter<Preference>(activity, 0, objects) {
        private val viewTypeCount = activity.viewTypeCount

        override fun getItemViewType(position: Int): Int {
            return getItem(position).viewType
        }

        override fun getViewTypeCount(): Int {
            return viewTypeCount
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val pref = getItem(position)

            val view: View = convertView ?: LayoutInflater.from(parent.context).inflate(pref!!.layout, parent, false)

            val title = view.findViewById<View>(android.R.id.title) as TextView
            title.setText(pref!!.title)

            if (pref is Item) {
                val icon = view.findViewById<View>(android.R.id.icon)
                if (icon != null) {
                    icon.visibility = View.GONE
                }

                val summary = view.findViewById<View>(android.R.id.summary) as TextView
                if (pref.summaryRes > 0) {
                    summary.setText(pref.summaryRes)
                } else {
                    summary.text = pref.summary
                }

                val widgetFrame = view.findViewById<View>(android.R.id.widget_frame) as ViewGroup
                if (pref.widget > 0) {
                    var widgetView: View? = widgetFrame.findViewById(android.R.id.checkbox)
                    if (widgetView == null) {
                        LayoutInflater.from(parent.context).inflate(pref.widget, widgetFrame)
                        widgetView = widgetFrame.findViewById(android.R.id.checkbox)
                    }
                    if (pref is ToggleItem) {
                        val checkBox: CompoundButton = widgetView as CompoundButton
                        checkBox.isChecked = pref.checked
                    }
                } else {
                    widgetFrame.visibility = View.GONE
                }
            }

            return view
        }

        override fun isEnabled(position: Int): Boolean {
            val pref = getItem(position)
            if (pref is Category) {
                return false
            }
            if (pref is Item) {
                return pref.enabled
            }
            return true
        }
    }

    override val layoutResource: Int
        get() = R.layout.activity_settings

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        refreshAnim.isInvisibleMode = true

        listView.emptyView = findViewById(android.R.id.empty)
        listView.adapter = preferenceAdapter
        listView.onItemClickListener = this
    }

    protected abstract fun createPreferenceItems(): List<Preference>
    protected abstract fun onPreferenceItemClick(action: Int, pref: Item)

    var isProgressVisible: Boolean = false
        set(value) {
            if (value) {
                refreshAnim.start()
            } else {
                refreshAnim.stop()
            }
        }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu
        menuInflater.inflate(R.menu.settings, menu)
        refreshAnim.menuItem = menu.findItem(R.id.menu_act_refresh)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val pref = listView.getItemAtPosition(position) as Preference
        if (pref is Item) {
            val action = pref.action
            if (pref is ToggleItem) {
                pref.toggle()
            }
            onPreferenceItemClick(action, pref)
        }
    }

    fun notifyDataSetChanged() {
        preferenceAdapter.notifyDataSetChanged()
    }
}
